### PR TITLE
feat: Show witnesses of non-exhaustiveness in `missing-match-arm` diagnostic

### PR DESCRIPTION
**Pull request by iDawer**
Shamelessly copied from rustc. Thus reporting format is same.

This extends public api  `hir::diagnostics::MissingMatchArms` with `uncovered_patterns: String` field. It does not expose data for implementing a quick fix yet.

-----
Worth to note: current implementation does not give a comprehensive list of missing patterns. Also mentioned in [paper](http://moscova.inria.fr/~maranget/papers/warn/warn.pdf):

> One may think that algorithm I should make an additional effort to provide more
> non-matching values, by systematically computing recursive calls on specialized
> matrices when possible, and by returning a list of all pattern vectors returned by
> recursive calls. We can first observe that it is not possible in general to supply the
> users with all non-matching values, since the signature of integers is (potentially)
> infinite.

### CODE REVIEW

File: crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs [hunk 1]
```diff
@@ -963,10 +981,67 @@ impl<'p> DeconstructedPat<'p> {
...
+                }
+            },
+            &Slice(Slice { _unimplemented: _void }) => unimplemented!(),
+            &Str(_void) => unimplemented!(),
+            &FloatRange(_void) => unimplemented!(),
+            IntRange(range) => return range.to_pat(cx, self.ty.clone()),
+            Wildcard | NonExhaustive => PatKind::Wild,
+            Missing { .. } => {
+                never!(
+                    "trying to convert a `Missing` constructor into a `Pat`; this is probably a bug,
```
Comment (flodiebold) [inline crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:1033]: when we're calling `never!`, it's *always* a bug ;)

File: crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs [hunk 2]
```diff
@@ -963,10 +981,67 @@ impl<'p> DeconstructedPat<'p> {
...
+                // ignore this issue.
+                TyKind::Ref(..) => PatKind::Deref { subpattern: subpatterns.next().unwrap() },
+                _ => {
+                    never!("unexpected ctor for type {:?} {:?}", self.ctor, self.ty);
+                    PatKind::Wild
+                }
+            },
+            &Slice(Slice { _unimplemented: _void }) => unimplemented!(),
+            &Str(_void) => unimplemented!(),
+            &FloatRange(_void) => unimplemented!(),
```
Comment (flodiebold) [inline crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:1028]: maybe use `never!` for these unimplementeds as well?

**Review by flodiebold — APPROVED** [review]
Looks good to me!

@bors delegate+

File: crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs (hunk 1 above)
Comment (iDawer) [reply crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:1033]: Definitely!

File: crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs (hunk 2 above)
Comment (iDawer) [reply crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:1028]: I understand these as 'to be implemented' not as bugs. As these variants are being implemented the `unimplemented!()`'s supposed to catch eyes.
They're (almost) all behind void enum, practically shouldn't ever hit them at runtime.

Comment (flodiebold) [reply crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:1028]: > practically shouldn't ever hit them at runtime

In other words, it's a bug if we ever hit them at runtime, right?

Can you instead `match void {}` to prove statically they won't be hit?

Comment (flodiebold) [reply crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:1028]: Just to clarify, `never!` being in the code is not a bug, it's completely normal. `never!` *being hit at runtime* is a bug.

Comment (iDawer) [reply crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs:1028]: > Can you instead `match void {}` to prove statically they won't be hit?

Yes, this is even better.

Comment (iDawer) [reply crates/hir-ty/src/diagnostics/match_check/deconstruct_pat.rs]: Done, thank you!
Good for merge?

**Comment by iDawer** [comment]
TIL: empty match expressions diverge.

Thank you for the review!

**Comment by lnicola** [comment]
image

**Comment by lnicola** [comment]
Is this supposed to trigger for something like

```rust
    match 20 {
        0..=10 => {}
    }
```

?

**Pull request merged**